### ISSUE TITLE
Scheduler drifts in cycles

### ISSUE THREAD

**Issue by b4dboi** [dagster-io/dagster#10881]
### Dagster version

1.1.2

### What's the issue?

We are currently running stress tests on dagster to find out what are the limits when running dagster on k8s. One of the test was designed to show how precise is the dagster scheduler. We started with 40 of concurrently schedules jobs (run 40 sleep jobs every minute) and then we exploited tables event_log and run_tags to evaluate different start up intervals. It showed up that even though each job should be fired at 00 seconds, as expected from the cron definition, the fire event is actually delayed by a few seconds. This delay is drifting over time in range 1-6 seconds from the expected scheduled time. The delay is increasing in approx. 18 minute cycles followed by a drop to 1 second. 

We distilled a minimal example with one sleep op, one job and one every-minute schedule and ran it on empty cluster, but the result is the same. There was no difference in setup with or without async thread pool, as you can see below (the first is sync, the second async). The blue line in both diagrams shows the difference between the scheduled time and the real time of job submission (calculated as PIPELINE_STARTING - .dagster/scheduled_execution_time). 

![mvp1per1minnothreads_latency_during_test](https://user-images.githubusercontent.com/11676900/205625221-083d97a9-5c3d-4776-9df9-f14e66d82434.png)

 
![mvp1per1min_latency_during_test](https://user-images.githubusercontent.com/11676900/205625233-6211a05e-5067-4993-8069-b503fcfee1e8.png)

The drift is notable even from the daemon log. Check the line at time 10:11:03.
```
dagster 2022-12-05 10:10:13 +0000 - dagster.daemon.SchedulerDaemon - INFO - No new tick times to evaluate for sleep_job_schedule
dagster 2022-12-05 10:10:43 +0000 - dagster.daemon.SensorDaemon - INFO - Not checking for any runs since no sensors have been started.
dagster 2022-12-05 10:11:03 +0000 - dagster.daemon.SchedulerDaemon - INFO - Evaluating schedule `sleep_job_schedule` at 2022-12-05 10:11:00 +0000
dagster 2022-12-05 10:11:04 +0000 - dagster.daemon.SchedulerDaemon - INFO - Completed scheduled launch of run 26be178c-4160-472c-81b2-e35db7b4a988 for sleep_job_schedule
dagster 2022-12-05 10:11:18 +0000 - dagster.daemon.SchedulerDaemon - INFO - Checking for new runs for the following schedules: sleep_job_schedule  
```

### What did you expect to happen?

We expect jobs to be executed on scheduled time.

### How to reproduce?

1. Deploy dagster using the official Helm chart with PostgreSQL database.
2. Create custom repository with following job and repository
```
import time
from collections import Counter

from dagster import job, op, repository, ScheduleDefinition, in_process_executor

@op
def sleep_op():
    time.sleep(1)

@job
def sleep_job():
    sleep_op()

sleep_schedule = ScheduleDefinition(job=sleep_job, cron_schedule="*/1 * * * *")

@repository(default_executor_def=in_process_executor)
def repo():
    return [sleep_schedule]

```
3. Deploy the user code deployment and activate the every-minute scheduler.
4. Monitor the drift in dagster-daemon log

### Deployment type

Dagster Helm chart

### Deployment details

We are using the official helm chart, but we have separated the user code deployment in the dedicated chart (https://artifacthub.io/packages/helm/dagster/dagster-user-deployments). Jobs are submitted via K8SRunLauncher without any tweaks. 

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by gibsondan** [comment]
Hi @b4dboi - I'm curious what exactly this latency number would have to be in order to be acceptable / what your use case is that would require those results.

It's interesting that the latency here is so cyclic and we can dig into why that would be, but if I'm reading these results correctly the latency to start a schedule is never higher than a few seconds for the benchmark that you're using. Generally the Dagster scheduler is designed to ensure that each schedule executes exactly once and that the latency doesn't increase dramatically as you add more schedules, rather than ensuring that every schedule is executed with, say, sub-second latency.

**Comment by gibsondan** [comment]
I'd also be curious if the cyclicness of the result here is the same when running the scheduler/daemon locally or if it's specifically something to do with running in k8s (I'd suspect that k8s doesn't affect this particular latency number)

**Comment by gibsondan** [comment]
I put out a change here that I believe will help with this specific benchmark at least, certainly in the case where there is a single schedule running: https://github.com/dagster-io/dagster/pull/10886/files

However, as the number of schedules in your Dagster instance increases, there would likely be some effect on latency as well. We have some levers that you can pull in the Helm chart to help with latency, like adding a threadpool that executes the schedules in parallel.

**Comment by b4dboi** [comment]
Let's say that the finding can negatively impact jobs where we expect precise scheduling because our source releases file every minute and we need to obtain that file as fast as possible. We could use sensors but we need to download the file/materialize it which consumes API requests anyways. From this perspective, downloading the file directly in job seems convenient to us.

The important part is "as fast as possible" and that is the reason why we run stress tests and digging into the each component of the start up time. As the scheduler submits the job in irregular intervals we cannot be sure what delay to expect in the end. I know this is true even for the rest of components in the process, but there is clearly something in the Scheduler implementation which affects the start-up delay at regular basis. I would expect some level of randomness at this component, but there is an obvious pattern, so I decided to submit an issue. I doubt this is an issue related to deployment in k8s, as monitoring the logs clearly shows that the scheduler is drifting (at least in our deployment).

We are trying to decompose the start-up time to better understand where are the limits and what are potential issues. Our goal is to minimize the start-up time and to make it as deterministic as possible. For example, adding the QueuedRunCordinator adds additional seconds to the start-up time. Although it's needed to make the deployment more robust the overall overhead is getting bigger and bigger in the stage we are not able to affect much. Just for the information we are aware of the negative impact of the K8S scheduler as you can see below (time between the job is created and the time pod is actually up and running).

![k8s_scheduler_delay](https://user-images.githubusercontent.com/11676900/205698730-9254023a-8e8f-476d-90c5-5deb370ad198.png)

**Comment by b4dboi** [comment]
> I put out a change here that I believe will help with this specific benchmark at least, certainly in the case where there is a single schedule running: https://github.com/dagster-io/dagster/pull/10886/files
> However, as the number of schedules in your Dagster instance increases, there would likely be some effect on latency as well. We have some levers that you can pull in the Helm chart to help with latency, like adding a threadpool that executes the schedules in parallel.

I'll check it, thank you. I think it is expectable that there will be effect on latency with increasing number of schedules. I was just surprised by the saw pattern which was totally independent of the number of schedules.

If you mean useThreads I already enabled this configuration and it had no effect on the result (actually it is the second diagram with titled mvp1per1min).

**Comment by gibsondan** [comment]
Thanks for the context - that makes sense. I think the attached PR should make the specific cyclic behavior that you observed more consistent once it lands.

**Comment by b4dboi** [comment]
I can confirm that the issue was fixed in the 1.1.7 as you can see below from the repeated test (blue line). I have some other observations regarding the scalability, but I think it's out of the scope for this issue. I will submit benchmark results into https://github.com/dagster-io/dagster/issues/7763.

Thank you for the fix. 

![test_mvp1_16threads_1min_latency_during_test](https://user-images.githubusercontent.com/11676900/209712379-e1af7c7f-8240-4ab8-89ed-ad2b9fe11aec.png)
